zipp.py

```diff
@@ -195,16 +193,18 @@ def joinpath(self, add):
 
     @staticmethod
     def _implied_dirs(names):
-        return more_itertools.unique_everseen(
-            parent + "/"
-            for name in names
-            for parent in _parents(name)
-            if parent + "/" not in names
-        )
+        seen = set(names)
```

Comment:
I see the rationale behind this change, but I also am saddened that what was once an elegant, single-expression on a generator expression is now several clauses with intertwined state. I'd like to mull this over a bit longer, but this is a good start for sure.

Reply:
I wonder if a better approach wouldn't be to use a simple set difference, e.g.

```py
parents = itertools.chain.from_iterable(map(_parents, names))
all_dirs = {p + posixpath.sep for p in parents}
return sorted(all_dirs - set(names))
```

Reply:
I'm disinclined to lose the order.

Reply:
```py
parents = itertools.chain.from_iterable(map(_parents, names))
# Cast names to a set for O(1) lookups
name_set = set(names)
# Deduplicate entries in original order
implied_dirs = OrderedDict.fromkeys(
    p + posixpath.sep for p in parents
    if p + posixpath.sep not in name_set
)
return iter(implied_dirs)
```

Reply:
I'm liking this last implementation quite a bit more.

Reply:
Let me know if you'd like me to make the change.

Reply:
Yes, please do.

Reply:
Would you like me to rebase or will you squash them before committing?

Reply:
> rebase or squash?

I often think it's useful to capture the evolution of a piece of code. It's the merge commit that represents the "acceptance" or "adoption" of the evolved branch. However, since GitHub captures that evolution, it's less important that it appears in the code repo history. Let's just leave it for now, if there's not anything in the history that you find particularly repugnant.

Reply:
No, I don't mind at all.